tests/components/tplink/test_init.py

```diff
@@ -209,13 +232,141 @@ async def test_platforms_are_initialized(hass):
+        )
+        data = hass.data[tplink.DOMAIN][COORDINATORS][switch.mac].data
+        assert data[CONF_HOST] == switch.host
+        assert data[CONF_MAC] == switch.sys_info["mac"]
+        assert data[CONF_MODEL] == switch.sys_info["model"]
+        assert data[CONF_SW_VERSION] == switch.sys_info["sw_ver"]
+        assert data[CONF_ALIAS] == switch.sys_info["alias"]
+        assert data[CONF_DEVICE_ID] == switch.sys_info["mac"]
+
+        emeter_readings = switch.get_emeter_realtime()
+        assert data[CONF_EMETER_PARAMS][ATTR_VOLTAGE] == round(
+            float(emeter_readings["voltage"]), 1
+        )
+        assert data[CONF_EMETER_PARAMS][ATTR_CURRENT_A] == round(
+            float(emeter_readings["current"]), 2
+        )
+        assert data[CONF_EMETER_PARAMS][ATTR_CURRENT_POWER_W] == round(
+            float(emeter_readings["power"]), 2
+        )
+        assert data[CONF_EMETER_PARAMS][ATTR_TOTAL_ENERGY_KWH] == round(
+            float(emeter_readings["total"]), 3
+        )
+        assert data[CONF_EMETER_PARAMS][ATTR_LAST_RESET][
+            ATTR_TOTAL_ENERGY_KWH
+        ] == utc_from_timestamp(0)
+
+        assert data[CONF_EMETER_PARAMS][ATTR_TODAY_ENERGY_KWH] == 1.123
+        assert data[CONF_EMETER_PARAMS][ATTR_LAST_RESET][
+            ATTR_TODAY_ENERGY_KWH
+        ] == datetime.utcnow().replace(hour=0, minute=0, second=0, microsecond=0)
+
+        assert discover.call_count == 0
+        assert get_static_devices.call_count == 1
+        assert light_setup.call_count == 1
+        assert switch_setup.call_count == 1
+
+
+async def test_smartplug_without_consumption_sensors(hass: HomeAssistant):
+    """Test that platforms are initialized per configuration array."""
+    config = {
+        tplink.DOMAIN: {
+            CONF_DISCOVERY: False,
+            CONF_SWITCH: [{CONF_HOST: "321.321.321.321"}],
+        }
+    }
+
+    with patch("homeassistant.components.tplink.common.Discover.discover"), patch(
+        "homeassistant.components.tplink.get_static_devices"
+    ) as get_static_devices, patch(
+        "homeassistant.components.tplink.common.SmartDevice._query_helper"
+    ), patch(
+        "homeassistant.components.tplink.light.async_setup_entry",
+        return_value=mock_coro(True),
+    ), patch(
+        "homeassistant.components.tplink.switch.async_setup_entry",
+        return_value=mock_coro(True),
+    ), patch(
+        "homeassistant.components.tplink.sensor.SmartPlugSensor.__init__"
+    ) as SmartPlugSensor, patch(
+        "homeassistant.components.tplink.common.SmartPlug.is_dimmable", False
+    ):
+
+        switch = SmartPlug("321.321.321.321")
+        switch.get_sysinfo = MagicMock(return_value=SMARTPLUGSWITCH_DATA["sysinfo"])
+        get_static_devices.return_value = SmartDevices([], [switch])
+
+        await async_setup_component(hass, tplink.DOMAIN, config)
+        await hass.async_block_till_done()
+
+        assert SmartPlugSensor.call_count == 0
```

Comment:
Check the precense or absence of this entity state in the core state machine instead.